new unnecessary_map_on_constructor lint

jonboh opened the pull request:
changelog: [`unnecessary_map_on_constructor`]: adds lint for cases in which map is not necessary. `Some(4).map(myfunction)` => `Some(myfunction(4))`

Closes https://github.com/rust-lang/rust-clippy/issues/6472

Note that the case mentioned in the issue `Some(..).and_then(|..| Some(..))` is fixed by a chain of lint changes. This PR completes the last part of that chain.

By `bind_instead_of_map`[lint](https://rust-lang.github.io/rust-clippy/master/index.html#/bind_instead_of_map):
`Some(4).and_then(|x| Some(foo(4)))` => `Some(4).map(|x| foo)`

By `redundant_closure` [lint](https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_closure):
`Some(4).map(|x| foo)` => `Some(4).map(fun)`

Finally by this PR `unnecessary_map_on_constructor`:
`Some(4).map(fun)` => `Some(fun(4))`

I'm not sure this is the desired behavior for clippy and if it should be addressed in another issue/PR. I'd be up to give it a try if that's the case.

bors commented:
:umbrella: The latest upstream changes (presumably #10692) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Alexendoo commented on tests/ui/unnecessary_map_on_constructor.rs:
```diff
@@ -0,0 +1,43 @@
+
+fn main() {
+    let x: i32 = 4;
+    //
+    let err = SimpleError {};
+    let a = Some(x); // => Some(fun(x))
+    let b: SimpleResult = Ok(x); // => Ok(fun(x))
+    let c: SimpleResult = Err(err); // => Ok(fun(x))
+    //
+    let a = Some(x).map(fun); // => Some(fun(x))
+    let b: SimpleResult = Ok(x).map(fun); // => Ok(fun(x))
+    let c: SimpleResult = Err(err).map_err(notfun); // => Ok(fun(x))
+    //
+    //
+    let a = Option::Some(x).map(fun); // => Some(fun(x))
+    let b: SimpleResult = SimpleResult::Ok(x).map(fun); // => SimpleResult::Ok(fun(x))
+    let c: SimpleResult = SimpleResult::Err(err).map_err(notfun); // => SimpleResult::Err(notfun(x))
+    let b: std::result::Result<i32, SimpleError> = Ok(x).map(fun); // => Ok(fun(x))
+    let c: std::result::Result<i32, SimpleError> = Err(err).map_err(notfun); // => Err(notfun(x))
+
+    a.map(fun); // these should not trigger
+    b.map(fun);
+    c.map_err(notfun);
+
+    let a = Some(fun(x));
+    let b: SimpleResult = Ok(fun(x));
+    let c: SimpleResult = Err(notfun(err)); // => Ok(fun(x))
```
The comments in the test can be dropped, some of them are incorrect and we'll be adding a standard way to denote it [in the near future](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Improve.20UI.20error.20checks)

jonboh replied:
[d4c2a9d](https://github.com/rust-lang/rust-clippy/pull/11413/commits/d4c2a9db436caa026f7d54123633da1da760ebcf)

Alexendoo commented on clippy_lints/src/unnecessary_map_on_constructor.rs:
```diff
@@ -0,0 +1,96 @@
+                return;
+            };
+
+            let (constructor_path, constructor_item) = if_chain! {
```
nit: for new stuff we use [let chains](https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md) rather than `if_chain`

jonboh replied:
[9ecab06](https://github.com/rust-lang/rust-clippy/pull/11413/commits/9ecab06f54aac5ba5fc82f6af99cc0908c5e7f3e)

Alexendoo commented on clippy_lints/src/unnecessary_map_on_constructor.rs:
```diff
@@ -0,0 +1,96 @@
+                        cx,
+                        UNNECESSARY_MAP_ON_CONSTRUCTOR,
+                        expr.span,
+                        &format!("unnecessary {fun_snippet} on contstuctor {constructor_snippet}(_)"),
+                        "try using",
```
```suggestion
                        &format!("unnecessary `{}` on constructor `{constructor_snippet}(_)`", path.ident.name),
                        "try",
```

jonboh replied:
I've removed the "using", but path.ident.name does not point to the `map` argument but to `map` itself, I've kept the `fun_snippet` for that reason

Alexendoo replied:
Yeah that's intended, the `map`/`map_err` is the unnecessary part, `fun_snippet` is necessary it's just moved around a bit

jonboh replied:
ok, I didn't realize that was for the warning message. After d501c3b
they and the solution proposed should be correct:

> warning: unnecessary map on constructor Ok(_)

> help: try: `Ok(fun(x))`

Alexendoo commented on clippy_lints/src/unnecessary_map_on_constructor.rs:
```diff
@@ -0,0 +1,96 @@
+                        expr.span,
+                        &format!("unnecessary {fun_snippet} on contstuctor {constructor_snippet}(_)"),
+                        "try using",
+                        suggestion,
```
nit: can inline `suggestion` here

jonboh replied:
[1cc4d0f](https://github.com/rust-lang/rust-clippy/pull/11413/commits/1cc4d0f43a8fd7e17febcac3b690115df5c53926)

Alexendoo commented on clippy_lints/src/unnecessary_map_on_constructor.rs:
```diff
@@ -0,0 +1,96 @@
+declare_clippy_lint! {
+    /// ### What it does
+    /// Suggest removing the use of a may (or map_err) method when an Option or Result is being construted.
+    /// ### Why is this bad?
```
A newline to make it nicer in `--explain`

```suggestion
    ///
    /// ### Why is this bad?
```

Alexendoo commented on clippy_lints/src/unnecessary_map_on_constructor.rs:
```diff
@@ -0,0 +1,96 @@
+    #[clippy::version = "1.73.0"]
+    pub UNNECESSARY_MAP_ON_CONSTRUCTOR,
+    complexity,
+    "Suggest removing usage of map in Option or Result construction"
```
```suggestion
    "using `map`/`map_err` on `Option` or `Result` constructors"
```

jonboh replied:
[4c15a60](https://github.com/rust-lang/rust-clippy/pull/11413/commits/4c15a6070b8a0686e1690412fbfd8670f3fcb11c)

Alexendoo reviewed (commented):
Thanks! Looks good, got a couple macro cases but mostly style nits

jonboh commented on tests/ui/unnecessary_map_on_constructor.rs:
yap, thanks for catching that
[d7cf575](https://github.com/rust-lang/rust-clippy/pull/11413/commits/d7cf5756d43d955f14fa836a4d7a66800c99aca6)

jonboh commented:
Thanks for the comments :), I think I've addressed all issues.

From the Zulip thread I get that the new syntax for the comments in the UI tests is still in the works, I've removed the comments, but if there's already some agreed upon syntax that I can add to be aligned with future practices let me know.

Alexendoo commented on tests/ui/unnecessary_map_on_constructor.rs:
```diff
@@ -0,0 +1,53 @@
+    let b: SimpleResult = Ok(fun(x));
+    let c: SimpleResult = Err(notfun(err));
+
+    // Should not trigger warning
```
would also be good to have tests that `Ok().map_err` and `Err().map` aren't linted

jonboh replied:
yes, should have thought about testing for that false positive, fixed on the squashed commit 
f136e16

Alexendoo commented on clippy_lints/src/unnecessary_map_on_constructor.rs:
```diff
@@ -0,0 +1,99 @@
+            return;
+        }
+        if let hir::ExprKind::MethodCall(path, recv, args, ..) = expr.kind {
+            let is_option = is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(recv), sym::Option);
+            let is_result = is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(recv), sym::Result);
```
Kind of a nit too but we can avoid a double lookup by checking if `get_type_diagnostic_name` matches `Some(sym::Option | sym::Result)`

Alexendoo reviewed (commented):
Thanks! just a couple last things I spotted and it looks good. If you could squash the commits too that would be great then it's ready for merging

Alexendoo commented:
Great, thanks!

@bors r+

bors commented:
:pushpin: Commit f136e1634a8d834d32deb89273717227eedaf153 has been approved by `Alexendoo`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=Alexendoo f136e1634a8d834d32deb89273717227eedaf153 -->
<!-- homu: {"type":"Approved","sha":"f136e1634a8d834d32deb89273717227eedaf153","approver":"Alexendoo","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit f136e1634a8d834d32deb89273717227eedaf153 with merge cb057019d46b823f142b6cc201319586e61471a8...
<!-- homu: {"type":"BuildStarted","head_sha":"f136e1634a8d834d32deb89273717227eedaf153","merge_sha":"cb057019d46b823f142b6cc201319586e61471a8"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/6162053211/job/16722729800), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/6162053212/job/16722749685), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/6162053208/job/16723062665)
Approved by: Alexendoo
Pushing cb057019d46b823f142b6cc201319586e61471a8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Alexendoo","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/6162053208/job/16723062665","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/6162053211/job/16722729800","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/6162053212/job/16722749685"},"merge_sha":"cb057019d46b823f142b6cc201319586e61471a8"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/6162053211/job/16722729800), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/6162053212/job/16722749685), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/6162053208/job/16723062665)
Approved by: Alexendoo
Pushing cb057019d46b823f142b6cc201319586e61471a8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Alexendoo","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/6162053208/job/16723062665","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/6162053211/job/16722729800","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/6162053212/job/16722749685"},"merge_sha":"cb057019d46b823f142b6cc201319586e61471a8"} -->
